chain/bitcoind_conn.go

```diff
@@ -139,10 +140,24 @@ func (c *BitcoindConn) Start() error {
 		return nil
 	}
 
+	var (
+		net wire.BitcoinNet
+		err error
+	)
+
 	// Verify that the node is running on the expected network.
-	net, err := c.getCurrentNet()
-	if err != nil {
-		return err
+	for {
+		net, err = c.getCurrentNet()
+		if err == nil {
+			// No error means bitcoind is responsive and we can proceed.
+			break
+		} else if err.(*btcjson.RPCError).Code == -28 {
+			// Error code -28 indicates the bitcoind rpc is warming up.
+			// We will wait and recheck periodically until rpc is ready.
+			time.Sleep(time.Second * 1)
```

Comment:
this looks okay, the only thing I'm worried about is bitcoind hanging, always returning this code. In that case it looks like we'll deadlock here.

Reply:
The alternative would be to crash, which is what happens now. This approach looks good to me, maybe just add a log like `Waiting for bitcoind RPC ready signal` that's only logged once if we ever run into this error code.

Reply:
There could perhaps be some upper limit of retries before returning an error. A couple of minutes perhaps? But yeah the alternative is crashing, and if ever bitcoind is stuck in this rpc warming up status, something is seriously wrong and requires intervention. I can't say I've ever seen bitcoind get stuck in this point, even if there's something wrong like db corruption it'll eventually crash in my experience.

Reply:
@sangaman in your experience, what has been the longest bitcoind has been in this warmup state? Waiting something like 5 mins sounds good to me.

Reply:
In my experience, even on a low-powered ARM machine this warm-up stage takes about a minute or two by default, rarely much more than than that. However, it's also possible to change the config to increase the amount of checks that are done before starting the bitcoind server, theoretically this could take hours if you had it do very deep/thorough checks.

Reply:
In that case, I'm fine with blocking indefinitely as long as a log statement exists.

Reply:
Agreed :)

Reply:
Yes, isn't bitcoind in that warmup state for the entire initial blockchain sync (which can take days, or more)?

I think it should wait indefinitely. However the sleep could maybe be increased slowly to a max of 10 minutes or so.

Reply:
No I don't think so, the warm up state is just while it's verifying the db and blocks on disk. Once it starts accepting and validating new blocks, it's past the warm up state and will respond to rpc requests normally.